src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/InteropServices/PlatformCompatabilityAnalyzer.cs

```diff
@@ -0,0 +1,446 @@
+                            }
+                        }
+                    }
+                }
+            }
+            return false;
+        }
+
+        private static void ReportDiagnosticsForAll(PooledDictionary<IOperation, ImmutableArray<PlatformAttributeInfo>> platformSpecificOperations, OperationBlockAnalysisContext context)
+        {
+            foreach (var platformSpecificOperation in platformSpecificOperations)
+            {
+                ReportDiagnostics(platformSpecificOperation.Key, platformSpecificOperation.Value.FirstOrDefault(), context);
+            }
+        }
+
+        private static void ReportDiagnostics(IOperation operation, PlatformAttributeInfo attribute, OperationBlockAnalysisContext context)
+        {
+            string operationName = string.Empty;
+
+            if (operation is IInvocationOperation iOperation)
+            {
+                operationName = iOperation.TargetMethod.Name;
+            }
+            else if (operation is IObjectCreationOperation cOperation)
+            {
+                operationName = cOperation.Constructor.Name;
+            }
+            else if (operation is IPropertyReferenceOperation pOperation)
+            {
+                operationName = pOperation.Property.Name;
+            }
+            else if (operation is IFieldReferenceOperation fOperation)
+            {
+                operationName = fOperation.Field.Name;
+            }
+            if (operation is IDelegateCreationOperation dOperation)
+            {
+                if (dOperation.Target is IMethodReferenceOperation mrOperation)
+                    operationName = mrOperation.Method.Name;
+            }
+            else if (operation is IEventReferenceOperation eOperation)
+            {
+                operationName = eOperation.Event.Name;
+            }
+
+            context.ReportDiagnostic(operation.CreateDiagnostic(SelectRule(attribute.AttributeType),
+                operationName, attribute.PlatformName, attribute.Version.ToString()));
+        }
+
+        private static void AnalyzeInvocationOperation(IOperation operation, OperationAnalysisContext context,
+            ref PooledDictionary<IOperation, ImmutableArray<PlatformAttributeInfo>> platformSpecificOperations)
+        {
+            using var builder = ArrayBuilder<PlatformAttributeInfo>.GetInstance();
+            AttributeData[]? attributes = null;
+
+            if (operation is IInvocationOperation iOperation)
+            {
+                attributes = FindAllPlatformAttributesApplied(iOperation.TargetMethod.GetAttributes(), iOperation.TargetMethod.ContainingType);
+            }
+            else if (operation is IObjectCreationOperation cOperation)
+            {
+                attributes = FindAllPlatformAttributesApplied(cOperation.Constructor.GetAttributes(), cOperation.Constructor.ContainingType);
+            }
+            else if (operation is IPropertyReferenceOperation pOperation)
```

Comment:
`IPropertyReferenceOperation`, `IFieldReferenceOperation`, `IMethodReferenceOperation` and `IEventReferenceOperation` all have a common base type `IMemberReferenceOperation` that exposes the underlying symbol. You can replace these 4 checks with a single type check.